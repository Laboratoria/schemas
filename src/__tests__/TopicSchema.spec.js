const mongoose = require('mongoose/browser');
const { TopicSchema } = require('../../')(mongoose);
const babelTopicJson = require('./fixtures/topics/babel');

describe('TopicSchema', () => {
  it('should fail validation when fields missing', () => {
    const doc = new mongoose.Document({}, TopicSchema);
    expect(doc.validateSync().errors).toMatchSnapshot();
  });

  it('should fail validation when syllabus is missing', () => {
    const doc = new mongoose.Document({
      slug: 'foo',
      repo: 'foo/bar',
      path: '.',
      version: '1.0.0',
      parserVersion: '2.0.0',
      track: 'js',
      locale: 'es-ES',
      title: 'Foo',
      stats: {
        partCount: 0,
        unitCount: 0,
        exerciseCount: 0,
        duration: 0,
        durationString: '0min',
      },
    }, TopicSchema);

    return doc.validate()
      .catch(err => expect(err.errors).toMatchSnapshot());
  });

  it('should fail validation when unit is missing parts', () => {
    const doc = new mongoose.Document({
      slug: 'foo',
      repo: 'foo/bar',
      path: '.',
      version: '1.0.0',
      parserVersion: '2.0.0',
      track: 'js',
      locale: 'es-ES',
      title: 'Foo',
      syllabus: {
        '01-intro': {
          order: 0,
          title: 'Intro to Foo',
          stats: {
            partCount: 4,
            exerciseCount: 0,
            duration: 35,
            durationString: '35min',
          },
        },
      },
      stats: {
        partCount: 0,
        unitCount: 0,
        exerciseCount: 0,
        duration: 0,
        durationString: '0min',
      },
    }, TopicSchema);

    return doc.validate()
      .catch(err => expect(err.errors['syllabus.01-intro.parts'].message).toBe('Path `parts` is required.'));
  });

  it('should fail validation when missing part body', () => {
    const doc = new mongoose.Document({
      slug: 'foo',
      repo: 'foo/bar',
      path: '.',
      version: '1.0.0',
      parserVersion: '2.0.0',
      track: 'js',
      locale: 'es-ES',
      title: 'Foo',
      syllabus: {
        '01-intro': {
          order: 0,
          title: 'Intro to Foo',
          parts: {
            '00-opening': {
              duration: 5,
              durationString: '5min',
              format: 'self-paced',
              type: 'read',
              title: 'Opening',
            },
            '01-quiz': {
              duration: 10,
              durationString: '10min',
              format: 'self-paced',
              type: 'quiz',
              title: 'Quiz 1',
            },
            '02-exercises': {
              duration: 10,
              durationString: '10min',
              format: 'self-paced',
              type: 'practice',
              title: 'Exercises',
            },
            '03-foo': {
              duration: 10,
              durationString: '10min',
              format: 'self-paced',
              type: 'read',
              title: 'Foo is bar',
            },
          },
          stats: {
            partCount: 4,
            exerciseCount: 0,
            duration: 35,
            durationString: '35min',
          },
        },
      },
      stats: {
        partCount: 4,
        unitCount: 1,
        exerciseCount: 0,
        duration: 35,
        durationString: '35min',
      },
    }, TopicSchema);

    return doc.validate()
      .catch(err => expect(err.errors).toMatchSnapshot());
  });

  it('should pass sync validation ...', () => {
    const doc = new mongoose.Document({
      slug: 'foo',
      repo: 'foo/bar',
      path: '.',
      version: '1.0.0',
      parserVersion: '2.0.0',
      track: 'js',
      locale: 'es-ES',
      title: 'Foo',
      syllabus: {},
      stats: {
        partCount: 0,
        unitCount: 0,
        exerciseCount: 0,
        durationString: '0min',
        duration: 0,
      },
    }, TopicSchema);
    expect(doc.validateSync()).toBe(undefined);
  });

  it('should validate existing topic json', () => {
    const babelTopic = new mongoose.Document(babelTopicJson, TopicSchema);
    return babelTopic.validate();
  });
});
