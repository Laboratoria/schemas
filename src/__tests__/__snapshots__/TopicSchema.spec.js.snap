// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicSchema should fail validation when fields missing 1`] = `
Object {
  "locale": [ValidatorError: Path \`locale\` is required.],
  "parserVersion": [ValidatorError: Path \`parserVersion\` is required.],
  "path": [ValidatorError: Path \`path\` is required.],
  "repo": [ValidatorError: Path \`repo\` is required.],
  "slug": [ValidatorError: Path \`slug\` is required.],
  "stats.duration": [ValidatorError: Path \`stats.duration\` is required.],
  "stats.durationString": [ValidatorError: Path \`stats.durationString\` is required.],
  "stats.exerciseCount": [ValidatorError: Path \`stats.exerciseCount\` is required.],
  "stats.partCount": [ValidatorError: Path \`stats.partCount\` is required.],
  "stats.unitCount": [ValidatorError: Path \`stats.unitCount\` is required.],
  "syllabus": [ValidatorError: Path \`syllabus\` is required.],
  "title": [ValidatorError: Path \`title\` is required.],
  "track": [ValidatorError: Path \`track\` is required.],
  "version": [ValidatorError: Path \`version\` is required.],
}
`;

exports[`TopicSchema should fail validation when missing part body 1`] = `
Object {
  "syllabus.01-intro.parts.00-opening": "Body is required for part type read",
  "syllabus.01-intro.parts.03-foo": "Body is required for part type read",
}
`;

exports[`TopicSchema should fail validation when syllabus is missing 1`] = `
Object {
  "syllabus": [ValidatorError: Path \`syllabus\` is required.],
}
`;
